### DISCUSSION TITLE
Why Isn't Directus Gaining More Attention in Mobile App Development?

### DISCUSSION THREAD

**Discussion by armanhadifatah** [directus/directus#22894] [Feature Requests]
### Summary

Directus may not receive as much attention in mobile app development as Firebase, Appwrite, and Supabase for several reasons:

### Basic Example

Directus stands out as a superior backend solution in many cases due to its flexibility, open-source nature, and robust feature set. Unlike Firebase, Appwrite, and Supabase, Directus provides a highly customizable headless CMS that can be tailored to specific project needs without being locked into a predefined ecosystem.

For Directus to become the go-to backend for mobile app development, a few enhancements could bridge the gap:

Flutter Package: Developing a dedicated package for Flutter would streamline integration, making it easier for developers to utilize Directus in their mobile apps.
Simple SSO Authentication: Implementing a straightforward Single Sign-On (SSO) solution would simplify user authentication processes, aligning Directus with the ease-of-use offered by competitors.

### Motivation

To make Directus a more attractive backend solution for mobile app developers.
Use Cases Supported:

Mobile app development with Flutter.
Simplified user authentication via SSO.
Expected Outcome:

Increased adoption of Directus in the mobile development community.
Enhanced developer experience and ease of integration with mobile apps.
Broader community engagement and contributions.
Constraints to Solve:

Lack of native support for Flutter.
Complexity in implementing SSO authentication.

### Detailed Design

Flutter Package:

Integration:

Create a Directus Flutter SDK.
Provide APIs for CRUD operations, real-time updates, and authentication.
Usage Example:

dart
Copy code
```
import 'package:directus_flutter/directus_flutter.dart';

final directus = Directus('https://your-directus-instance.com');
var data = await directus.items('your_collection').read();
Corner Cases:
```

Handle network failures gracefully.
Ensure data synchronization in offline mode.
SSO Authentication:

Integration:

Implement OAuth2 and OpenID Connect support.
Provide configuration options for popular SSO providers (Google, Facebook, etc.).
Usage Example:

javascript
Copy code
// Client-side SSO login
`directus.auth.loginWithSSO('google');`
Corner Cases:

Manage token refresh and expiration.
Provide fallback for SSO provider downtime.

### Requirements List

MoSCoW Method:

Must-Have:

Develop a dedicated Flutter package for seamless Directus integration.
Implement simple SSO authentication.
Should-Have:

Comprehensive documentation and migration guides.
Community support and feedback channels.
Could-Have:

Example projects demonstrating integration with Flutter and SSO.
Additional plugins for extended functionalities.
Won't-Have:

Major overhauls to existing Directus core functionalities.

### Drawbacks

Drawbacks
Implementation Cost:

Significant development effort required, increasing code size and complexity.
User Space Implementation:

The proposed features could be partially implemented through custom user solutions, though less seamlessly.
Impact on Teaching:

New features add complexity, potentially complicating the learning curve for new users.
Integration Challenges:

Ensuring compatibility with existing and planned Directus features may be complex and require careful design.
Migration Costs:

Although designed to be non-breaking, existing applications might still need adjustments, incurring some migration costs.

### Alternatives

Considered Designs:

Third-Party Integrations: Using existing third-party solutions for Flutter integration and SSO authentication, though this may lack seamless integration and control.
Custom Plugins: Allowing the community to develop custom plugins for Flutter and SSO, which may lead to inconsistent quality and support.
Impact of Not Implementing:

Missed Opportunities: Without these enhancements, Directus may lose potential mobile app developers to more integrated solutions like Firebase, Appwrite, and Supabase.
Limited Growth: The Directus ecosystem may experience slower growth and fewer contributions from the mobile development community.

### Adoption Strategy

Adoption Strategy
Implementation and Adoption by Existing Developers:

Non-Breaking Changes:

The new Flutter package and simple SSO authentication will be optional add-ons, ensuring current functionalities remain intact.
Smooth Migration Path:

Comprehensive migration guides and tools will facilitate a seamless transition, with detailed documentation and example code.
Impact on the Directus Ecosystem:

Enhanced Flexibility: The Flutter package will attract more mobile developers, fostering contributions and extensions.
Improved Authentication: Simplified SSO will enhance security and user management across all projects.
Broader Adoption: These enhancements will increase Directus's competitiveness and community growth.

**Comment by NilsBaumgartner1994** [comment]
Have a look at our PKCE solution?: https://github.com/directus/directus/discussions/23611
